File: server/routes/lists.js
const express = require('express');
const router = express.Router();
const passport = require('passport');
const { body, validationResult } = require('express-validator');
const List = require('../models/List');
const Task = require('../models/Task');

const auth = passport.authenticate('jwt', { session: false });

// Get all lists for user
router.get('/', auth, async (req, res) => {
  try {
    const lists = await List.find({
      user: req.user._id,
      isArchived: false
    }).sort({ order: 1, createdAt: 1 });
    
    // Get task counts for each list
    const listsWithCounts = await Promise.all(
      lists.map(async (list) => {
        const taskCount = await Task.countDocuments({
          list: list._id,
          status: 'active'
        });
        
        return {
          ...list.toObject(),
          taskCount
        };
      })
    );
    
    res.json({ lists: listsWithCounts });
  } catch (error) {
    console.error('Get lists error:', error);
    res.status(500).json({ error: { message: 'Failed to fetch lists' } });
  }
});

// Get single list
router.get('/:id', auth, async (req, res) => {
  try {
    const list = await List.findOne({
      _id: req.params.id,
      user: req.user._id
    });
    
    if (!list) {
      return res.status(404).json({ error: { message: 'List not found' } });
    }
    
    res.json({ list });
  } catch (error) {
    console.error('Get list error:', error);
    res.status(500).json({ error: { message: 'Failed to fetch list' } });
  }
});

// Create list
router.post('/',
  auth,
  [
    body('name').trim().notEmpty()
  ],
  async (req, res) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ errors: errors.array() });
      }
      
      const list = new List({
        ...req.body,
        user: req.user._id
      });
      
      await list.save();
      
      // Emit socket event
      const io = req.app.get('io');
      io.to(`user:${req.user._id}`).emit('list:created', list);
      
      res.status(201).json({ list });
    } catch (error) {
      console.error('Create list error:', error);
      res.status(500).json({ error: { message: 'Failed to create list' } });
    }
  }
);

// Update list
router.put('/:id', auth, async (req, res) => {
  try {
    const list = await List.findOne({
      _id: req.params.id,
      user: req.user._id
    });
    
    if (!list) {
      return res.status(404).json({ error: { message: 'List not found' } });
    }
    
    // Prevent updating default lists' core properties
    if (list.isDefault) {
      const allowedUpdates = ['color', 'order'];
      Object.keys(req.body).forEach(key => {
        if (allowedUpdates.includes(key)) {
          list[key] = req.body[key];
        }
      });
    } else {
      Object.keys(req.body).forEach(key => {
        if (key !== '_id' && key !== 'user') {
          list[key] = req.body[key];
        }
      });
    }
    
    await list.save();
    
    // Emit socket event
    const io = req.app.get('io');
    io.to(`user:${req.user._id}`).emit('list:updated', list);
    
    res.json({ list });
  } catch (error) {
    console.error('Update list error:', error);
    res.status(500).json({ error: { message: 'Failed to update list' } });
  }
});

// Delete list
router.delete('/:id', auth, async (req, res) => {
  try {
    const list = await List.findOne({
      _id: req.params.id,
      user: req.user._id
    });
    
    if (!list) {
      return res.status(404).json({ error: { message: 'List not found' } });
    }
    
    // Prevent deleting default lists
    if (list.isDefault) {
      return res.status(400).json({ error: { message: 'Cannot delete default lists' } });
    }
    
    // Move tasks to inbox
    const inboxList = await List.findOne({
      user: req.user._id,
      name: 'Inbox'
    });
    
    if (inboxList) {
      await Task.updateMany(
        { list: list._id },
        { $set: { list: inboxList._id } }
      );
    }
    
    await list.deleteOne();
    
    // Emit socket event
    const io = req.app.get('io');
    io.to(`user:${req.user._id}`).emit('list:deleted', { id: list._id });
    
    res.json({ message: 'List deleted successfully' });
  } catch (error) {
    console.error('Delete list error:', error);
    res.status(500).json({ error: { message: 'Failed to delete list' } });
  }
});

// Reorder lists
router.patch('/reorder', auth, async (req, res) => {
  try {
    const { orders } = req.body; // Array of { id, order }
    
    if (!Array.isArray(orders)) {
      return res.status(400).json({ error: { message: 'Invalid orders format' } });
    }
    
    const bulkOps = orders.map(item => ({
      updateOne: {
        filter: { _id: item.id, user: req.user._id },
        update: { $set: { order: item.order } }
      }
    }));
    
    await List.bulkWrite(bulkOps);
    
    res.json({ message: 'Lists reordered successfully' });
  } catch (error) {
    console.error('Reorder lists error:', error);
    res.status(500).json({ error: { message: 'Failed to reorder lists' } });
  }
});

module.exports = router;

File: app/routes/lists.py
"""
List routes
"""
from fastapi import APIRouter, HTTPException, Depends, status
from app.models import ListCreate, ListUpdate
from app.auth import get_current_user
from app.database import get_database
from bson import ObjectId

router = APIRouter()


@router.get("/", response_model=dict)
async def get_lists(current_user: dict = Depends(get_current_user)):
    """Get all lists for the user"""
    db = get_database()
    
    lists = await db.lists.find({
        "user": ObjectId(current_user["_id"]),
        "isArchived": False
    }).sort("order", 1).to_list(None)
    
    # Get task counts - use original ObjectId before converting
    for list_item in lists:
        task_count = await db.tasks.count_documents({
            "list": list_item["_id"],  # Use original ObjectId
            "status": {"$nin": ["completed", "skipped"]}  # Count non-completed tasks
        })
        list_item["taskCount"] = task_count
        # Now convert IDs to strings
        list_item["_id"] = str(list_item["_id"])
        list_item["user"] = str(list_item["user"])
    
    return {"lists": lists}


@router.post("/", response_model=dict, status_code=status.HTTP_201_CREATED)
async def create_list(list_data: ListCreate, current_user: dict = Depends(get_current_user)):
    """Create a new list"""
    db = get_database()
    
    list_doc = {
        "name": list_data.name,
        "user": ObjectId(current_user["_id"]),
        "color": list_data.color,
        "icon": list_data.icon,
        "isDefault": False,
        "isSmart": False,
        "smartFilter": None,
        "parent": ObjectId(list_data.parent) if list_data.parent else None,
        "order": 0,
        "isArchived": False
    }
    
    result = await db.lists.insert_one(list_doc)
    list_id = result.inserted_id
    
    list_doc["_id"] = str(list_id)
    list_doc["user"] = str(list_doc["user"])
    list_doc["taskCount"] = 0
    
    return {"list": list_doc}


@router.put("/{list_id}", response_model=dict)
async def update_list(
    list_id: str,
    list_data: ListUpdate,
    current_user: dict = Depends(get_current_user)
):
    """Update a list"""
    db = get_database()
    
    list_item = await db.lists.find_one({
        "_id": ObjectId(list_id),
        "user": ObjectId(current_user["_id"])
    })
    
    if not list_item:
        raise HTTPException(status_code=404, detail="List not found")
    
    update_data = {}
    if list_data.name is not None:
        update_data["name"] = list_data.name
    if list_data.color is not None:
        update_data["color"] = list_data.color
    if list_data.icon is not None:
        update_data["icon"] = list_data.icon
    if list_data.order is not None:
        update_data["order"] = list_data.order
    
    if update_data:
        await db.lists.update_one(
            {"_id": ObjectId(list_id)},
            {"$set": update_data}
        )
    
    updated_list = await db.lists.find_one({"_id": ObjectId(list_id)})
    updated_list["_id"] = str(updated_list["_id"])
    updated_list["user"] = str(updated_list["user"])
    
    return {"list": updated_list}


@router.delete("/{list_id}", response_model=dict)
async def delete_list(list_id: str, current_user: dict = Depends(get_current_user)):
    """Delete a list"""
    db = get_database()
    
    list_item = await db.lists.find_one({
        "_id": ObjectId(list_id),
        "user": ObjectId(current_user["_id"])
    })
    
    if not list_item:
        raise HTTPException(status_code=404, detail="List not found")
    
    if list_item.get("isDefault"):
        raise HTTPException(status_code=400, detail="Cannot delete default lists")
    
    # Move tasks to inbox
    inbox = await db.lists.find_one({
        "user": ObjectId(current_user["_id"]),
        "name": "Inbox"
    })
    
    if inbox:
        await db.tasks.update_many(
            {"list": ObjectId(list_id)},
            {"$set": {"list": inbox["_id"]}}
        )
    
    await db.lists.delete_one({"_id": ObjectId(list_id)})
    
    return {"message": "List deleted successfully"}

